scipy/stats/tests/test_sampling.py

```diff
@@ -0,0 +1,619 @@
+            self.shift = shift
+
+        def pdf(self, x):
+            x -= self.shift
+            y = 1. / (abs(x) + 1.)
+            return 0.5 * y * y
+
+        def dpdf(self, x):
+            x -= self.shift
+            y = 1. / (abs(x) + 1.)
+            y = y * y * y
+            return y if (x < 0.) else -y
+
+        def cdf(self, x):
+            x -= self.shift
+            if x <= 0.:
+                return 0.5 / (1. - x)
+            else:
+                return 1. - 0.5 / (1. + x)
+
+    dists = [dist0(), dist1(), dist2(0.), dist2(10000.)]
+
+    mv0 = [0., 4./15.]
+    mv1 = [0., 0.01]
+    mv2 = [0., np.inf]
+    mv3 = [10000., np.inf]
+    mvs = [mv0, mv1, mv2, mv3]
+
+    @pytest.mark.parametrize("dist, mv_ex",
+                             zip(dists, mvs))
+    def test_basic(self, dist, mv_ex):
+        with suppress_warnings() as sup:
+            # filter the warnings thrown by UNU.RAN
+            sup.filter(RuntimeWarning)
+            rng = TransformedDensityRejection(dist, seed=42)
+        rvs = rng.rvs(100000)
+        mv = rvs.mean(), rvs.var()
+        # test the moments only if the variance is finite
+        if np.isfinite(mv_ex[1]):
+            assert_allclose(mv, mv_ex, rtol=1e-7, atol=1e-1)
+        # Cramer Von Mises test for goodness-of-fit
+        rvs = rng.rvs(500)
+        dist.cdf = np.vectorize(dist.cdf)
+        pval = cramervonmises(rvs, dist.cdf).pvalue
+        assert pval > 0.1
+
+    # PDF 0 everywhere => bad construction points
+    bad_pdfs = [(lambda x: 0, UNURANError, r"50 : bad construction points.")]
+    bad_pdfs += bad_pdfs_common  # type: ignore[arg-type]
+
+    @pytest.mark.parametrize("pdf, err, msg", bad_pdfs)
+    def test_bad_pdf(self, pdf, err, msg):
+        class dist:
+            pass
+        dist.pdf = pdf
+        dist.dpdf = lambda x: 1  # an arbitrary dPDF
+        with pytest.raises(err, match=msg):
+            TransformedDensityRejection(dist)
+
+    @pytest.mark.parametrize("dpdf, err, msg", bad_dpdf_common)
+    def test_bad_dpdf(self, dpdf, err, msg):
+        class dist:
+            pass
+        dist.pdf = lambda x: x
+        dist.dpdf = dpdf
+        with pytest.raises(err, match=msg):
+            TransformedDensityRejection(dist, domain=(1, 10))
+
+    # test domains with inf + nan in them. need to write a custom test for
+    # this because not all methods support infinite tails.
+    @pytest.mark.parametrize("domain, err, msg", inf_nan_domains)
+    def test_inf_nan_domains(self, domain, err, msg):
+        with pytest.raises(err, match=msg):
+            TransformedDensityRejection(common_cont_dist, domain=domain)
+
+    # TODO: for cpoints < 0, UNU.RAN throws a warning and sets cpoints
+    #       to a default value. This is not consistent with other invalid
+    #       values of cpoints for which it throws errors. Hence, We should
+    #       validate this parameter ourself.
+    @pytest.mark.parametrize("cpoints", [0, 0.1])
```

Comment:
> i am confused :) why is an error raised in that case? and why is cpoints < 0 invalid?

This is an inconsistency in UNU.RAN. That's why I have a TODO comment above it. To throw an error when `cpoints` is negative or non-integer, we need to validate the argument ourselves. I will do that in the next commit.

Reply:
done.

Reply:
In the tutorial, we say:

> This can be changed by passing a ``cpoints`` parameter which can either be an array of construction points or an integer representing the number of construction points to use.

In the docstring, we say 

> cpoints : int, optional. The number of construction points. Default is 30.

I think that is why I misunderstood the test. in the docstring, should we say:
```
cpoints : int or array-like, optional. 
    If cpoints is an interger, it defines the number of construction points. If it is array-like, the elements are the construction points. Default is 30.
```
?

Reply:
You are right! I will add this in the documentation, thanks!

Reply:
done.